runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,219 @@
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+
+import org.corfudb.protocols.logprotocol.MultiObjectSMREntry;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.SubscriberAlreadyExistsException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.ObjectsView;
+import org.corfudb.runtime.view.StreamOptions;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.runtime.view.stream.IStreamView;
+
+/**
+ *
+ * Created by hisundar on 04/28/2020.
+ */
+@Slf4j
+public class StreamManager {
+
+    private final int MAX_SUBSCRIBERS = 12; // Arbitrary sanity check value.
+
+    /**
+     * Corfu Runtime.
+     */
+    @Getter
+    private final CorfuRuntime runtime;
+
+    /**
+     * Map of StreamingSubscriptionThreads.
+     * <p>
+     * This is the map of all streaming subscriptions.
+     */
+    @Getter
+    private final Map<StreamListener, SubscriptionThread> subscriptions = new HashMap<>();
+
+    public StreamManager(@Nonnull CorfuRuntime runtime) {
+        this.runtime = runtime;
+    }
+
+    /**
+     * Subscribe to updates.
+     *
+     * @param streamListener   Client listener.
+     * @param namespace        Namespace of interest.
+     * @param tablesOfInterest Only updates from these tables will be returned.
+     * @param startAddress     Address to start the notifications from.
+     */
+    public synchronized <K extends Message, V extends Message, M extends Message>
+    void subscribe(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                   @Nonnull List<TableSchema> tablesOfInterest, long startAddress) {
+        if (subscriptions.size() >= MAX_SUBSCRIBERS) {
+            throw new SubscriberAlreadyExistsException(
+                    "subscribe: Max "+MAX_SUBSCRIBERS+" hit! Consider consolidating "+streamListener);
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new SubscriberAlreadyExistsException(
+                    "StreamManager:subscrbie already registered "+streamListener+". Maybe call unregister");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}", streamListener,
+                startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public synchronized void unsubscribe(@Nonnull StreamListener streamListener) {
+        SubscriptionThread st = subscriptions.remove(streamListener);
+        if (st == null) {
+            log.warn("StreamManager::unsubscribe has no context for "+streamListener.toString());
```

Comment:
NIT: When concating the strings I think its better to have spaces around `+`

Reply:
done. thanks

Reply:
> Looks like this is related to the stream tagging feature? I see that the StreamManager uses the txn stream but filters out based on the tables of interest. Do you plan to change it once stream tagging is available?

Yes as soon as your change to have stream tagging is merged we just have to modify the TRANSACTION_STREAM to that specific stream tag. Which is exactly why this change is needed in the first place. In the previous (current) implementation there is just one publisher  which means we can't have selective polling.

Reply:
I would vote for `log.warn("StreamManager::unsubscribe has no context for {}", streamListener);`

At least it won't create an additional instance of a String, and appears prettier